dvc/remote/base.py

```diff
@@ -667,16 +667,28 @@ def _checkout_file(
         self, path_info, checksum, force, progress_callback=None
     ):
         cache_info = self.checksum_to_path_info(checksum)
-        if self.exists(path_info):
-            msg = "data '{}' exists. Removing before checkout."
-            logger.warning(msg.format(str(path_info)))
-            self.safe_remove(path_info, force=force)
 
-        self.link(cache_info, path_info)
-        self.state.save_link(path_info)
-        self.state.save(path_info, checksum)
-        if progress_callback:
-            progress_callback.update(str(path_info))
+        identical_exists = False
+        if self.exists(path_info):
+            if (
+                self.state.get(path_info) == checksum
+                and not force
+                # NOTE In case of protected repo, we need to overwrite even
+                # identical file to make it protected again
+                and not self.protected
```

Comment:
do we make it unprotected somewhere?

Reply:
In RemoteLOCAL, yes.

Reply:
could you give a little bit more details, please? Where does it happen? The logic looks weird tbh - that we are unprotecting everything somewhere outside and then protecting one by one here again. Why don't we do unprotect here one by one as well?

Reply:
Ok, let's try to review the use case this change tries to tackle.
Here is code that can reproduce the behavior:
```def test_metric(dvc_repo, repo_dir):
    dvc_repo.add(repo_dir.FOO)
    dvc_repo.add(repo_dir.BAR)

    shutil.copy(repo_dir.BAR, repo_dir.FOO)

    dvc_repo.add(repo_dir.FOO)
 ```
This use case is quite specific and assumes modifing existing, and `dvc` controlled file to other file that content is known and cached by `dvc`.

What has been reported in original case, was that `dvc` tries to remove the file and check out "the new content." It is ok, but it would be better if `dvc` would just change checksum in stage file of modified file. 

Easiest way to handle that is just to get checksum for path from state. If checksum got from state database matches modified file checksum, there is no need to remove it and check it out. (We want to do it in case of force) . Probably `identical_exists` is unfortunate name, it should rather be `already_exists`.

Now, why we have special if for `self.protected`? When we have protected repo (example test is `test_add.py/TestAddUnprotected`, after "re-adding" target that was unprotected we would like it to have properties as other protected files. If we won't remove this file and check it out, it will stay unprotected.  So I can either mark it as "already_exists" and handle unprotecting by myself, or just remove the file and let original logic handle `protected` flag.

Reply:
>  If checksum got from state database matches modified file checksum

what are cases we will be able to get the checksum for the modified file from state db? what are cases when not?

>  after "re-adding" target that was unprotected

where do we check if it was unprotected? We rely on a global repo setting for all cases which can suboptimal as far as I understand. Would be great to understand better again how will this logic behave for `dvc run`, `dvc repro`, `dvc add` etc for different scenarios, how will it look like from the UI perspective.

Reply:
> what are cases we will be able to get the checksum for the modified file from state db? what are cases when not?

From db or from cache? From db, checksum will be avialable after any operation that touches particular file. In given example, db has "modified" checksum, because before adding, there is `status` performed and it updates state for this file. 

> where do we check if it was unprotected? 

We do not, thats the point. If cache is protected, we know that target can be unprotected or protected, allowing to remove it and relink from cache makes sure that it will be added in protected way, which makes sense, because from user point of view this file should be treated as "new", even though it is not.

I agree that relaying on global setting might not be optimal, but here the problem lies in `scope` of protection. Whole cache can be `protected` but only file can be `unportected`. (For convinience I omit case where we unprotect all files inside repo). I don't see any obvious way to check if file is protected, besides some access permissions, which can also be dependant on permissions of user in particular system.